[Bug]: [storagev2]streaming node panic during test after milvus recovered from many chaos test

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Environment

```markdown
- Milvus version:master-20250701-c381bf3e-amd64
- Deployment mode(standalone or cluster):
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

```
[2025/07/02 03:18:35.599 +00:00] [DEBUG] [recovery/recovery_storage_impl.go:186] ["skip the message before the checkpoint"] [module=streamingnode] [component=recovery-storage] [channel=by-dev-rootcoord-dml_9:rw@11] [state=working] [message="{type=TIME_TICK,vchannel=timetick=459125816699387905,messageID=21/582/0,lastConfirmed=21/578/0,size=17}"] [checkpoint=459125848576107315] [incoming=459125816699387905]
[2025/07/02 03:18:35.599 +00:00] [DEBUG] [recovery/recovery_storage_impl.go:186] ["skip the message before the checkpoint"] [module=streamingnode] [component=recovery-storage] [channel=by-dev-rootcoord-dml_9:rw@11] [state=working] [message="{type=DELETE,vchannel=by-dev-rootcoord-dml_9_459125116622022930v0,timetick=459125817695797257,messageID=21/583/0,lastConfirmed=21/581/0,size=24081,collectionID=459125116622022930,rows=1496}"] [checkpoint=459125848576107315] [incoming=459125817695797257]
[2025/07/02 03:18:35.599 +00:00] [DEBUG] [pipeline/flow_graph_dd_node.go:224] ["DDNode receive delete messages"] [channel=by-dev-rootcoord-dml_9_459125116622022930v0] [numRows=1496] [startPosTs=459125817695797257] [endPosTs=459125817695797257]
[2025/07/02 03:18:35.599 +00:00] [DEBUG] [recovery/recovery_storage_impl.go:186] ["skip the message before the checkpoint"] [module=streamingnode] [component=recovery-storage] [channel=by-dev-rootcoord-dml_9:rw@11] [state=working] [message="{type=TIME_TICK,vchannel=timetick=459125817695797258,messageID=21/584/0,lastConfirmed=21/581/0,size=17}"] [checkpoint=459125848576107315] [incoming=459125817695797258]
github.com/milvus-io/milvus/internal/storage.BuildRecord
	/workspace/source/internal/storage/serde.go:895 pc=0x3fcaf40


[2025/07/02 03:18:36.808 +00:00] [ERROR] [conc/options.go:54] ["Conc pool panicked"] [panic="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/milvus-io/milvus/pkg/v2/util/conc.(*poolOption).antsOptions.func1\n\t/workspace/source/pkg/util/conc/options.go:54\ngithub.com/panjf2000/ants/v2.(*goWorker).run.func1.1\n\t/go/pkg/mod/github.com/panjf2000/ants/v2@v2.11.3/worker.go:60\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:792\ngithub.com/milvus-io/milvus/pkg/v2/util/conc.(*Pool[...]).Submit.func1.1\n\t/workspace/source/pkg/util/conc/pool.go:75\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:792\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:262\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:925\ngithub.com/milvus-io/milvus/internal/storage.BuildRecord\n\t/workspace/source/internal/storage/serde.go:895\ngithub.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).serializeBinlog\n\t/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:175\ngithub.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).writeInserts\n\t/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:107\ngithub.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).Write\n\t/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:79\ngithub.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*SyncTask).Run\n\t/workspace/source/internal/flushcommon/syncmgr/task.go:132\ngithub.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*keyLockDispatcher[...]).Submit.func1\n\t/workspace/source/internal/flushcommon/syncmgr/key_lock_dispatcher.go:39\ngithub.com/milvus-io/milvus/pkg/v2/util/conc.(*Pool[...]).Submit.func1\n\t/workspace/source/pkg/util/conc/pool.go:82\ngithub.com/panjf2000/ants/v2.(*goWorker).run.func1\n\t/go/pkg/mod/github.com/panjf2000/ants/v2@v2.11.3/worker.go:73"]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0x3fcaf40]

goroutine 48830 gp=0xc0025af340 m=366 mp=0xc00277b808 [running]:
panic({0x5a05f00?, 0x90d72b0?})
	/usr/local/go/src/runtime/panic.go:811 +0x168 fp=0xc001eef4b8 sp=0xc001eef408 pc=0x1f6a828
github.com/milvus-io/milvus/pkg/v2/util/conc.(*poolOption).antsOptions.func1({0x5a05f00, 0x90d72b0})
	/workspace/source/pkg/util/conc/options.go:56 +0x146 fp=0xc001eef580 sp=0xc001eef4b8 pc=0x3ed9746
github.com/panjf2000/ants/v2.(*goWorker).run.func1.1()
	/go/pkg/mod/github.com/panjf2000/ants/v2@v2.11.3/worker.go:60 +0xd0 fp=0xc001eef608 sp=0xc001eef580 pc=0x3ed6170
panic({0x5a05f00?, 0x90d72b0?})
	/usr/local/go/src/runtime/panic.go:792 +0x132 fp=0xc001eef6b8 sp=0xc001eef608 pc=0x1f6a7f2
github.com/milvus-io/milvus/pkg/v2/util/conc.(*Pool[...]).Submit.func1.1()
	/workspace/source/pkg/util/conc/pool.go:75 +0x8a fp=0xc001eef718 sp=0xc001eef6b8 pc=0x437012a
panic({0x5a05f00?, 0x90d72b0?})
	/usr/local/go/src/runtime/panic.go:792 +0x132 fp=0xc001eef7c8 sp=0xc001eef718 pc=0x1f6a7f2
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:262
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:925 +0x359 fp=0xc001eef828 sp=0xc001eef7c8 pc=0x1f6dad9
github.com/milvus-io/milvus/internal/storage.BuildRecord(0xc009605d40, 0xc002c0e4e0, {0xc00ae4e700, 0x1c, 0xc00a1ed520?})
	/workspace/source/internal/storage/serde.go:895 +0x100 fp=0xc001eef8a0 sp=0xc001eef828 pc=0x3fcaf40
github.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).serializeBinlog(0xc001eefe30, {0x1c?, 0x7fd2d2cc4438?}, 0xc001fb0300)
	/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:175 +0x192 fp=0xc001eef968 sp=0xc001eef8a0 pc=0x436a752
github.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).writeInserts(0xc001eefe30, {0x691fe80, 0x93e53e0}, 0xc001fb0300)
	/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:107 +0xc5 fp=0xc001eefab0 sp=0xc001eef968 pc=0x4369f05
github.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*BulkPackWriterV2).Write(0xc001eefe30, {0x691fe80, 0x93e53e0}, 0xc001fb0300)
	/workspace/source/internal/flushcommon/syncmgr/pack_writer_v2.go:79 +0x1b3 fp=0xc001eefbe0 sp=0xc001eefab0 pc=0x43697d3
github.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*SyncTask).Run(0xc002a3c000, {0x691fe80, 0x93e53e0})
	/workspace/source/internal/flushcommon/syncmgr/task.go:132 +0x2dd fp=0xc001eefea0 sp=0xc001eefbe0 pc=0x436dfdd
github.com/milvus-io/milvus/internal/flushcommon/syncmgr.(*keyLockDispatcher[...]).Submit.func1()
	/workspace/source/internal/flushcommon/syncmgr/key_lock_dispatcher.go:39 +0x88 fp=0xc001eeff20 sp=0xc001eefea0 pc=0x43703a8
github.com/milvus-io/milvus/pkg/v2/util/conc.(*Pool[...]).Submit.func1()
	/workspace/source/pkg/util/conc/pool.go:82 +0x9e fp=0xc001eeff80 sp=0xc001eeff20 pc=0x437001e
github.com/panjf2000/ants/v2.(*goWorker).run.func1()
	/go/pkg/mod/github.com/panjf2000/ants/v2@v2.11.3/worker.go:73 +0x8a fp=0xc001eeffe0 sp=0xc001eeff80 pc=0x3ed602a
runtime.goexit({})
	/usr/local/go/src/runtime/asm_amd64.s:1700 +0x1 fp=0xc001eeffe8 sp=0xc001eeffe0 pc=0x1f74261
created by github.com/panjf2000/ants/v2.(*goWorker).run in goroutine 48776
	/go/pkg/mod/github.com/panjf2000/ants/v2@v2.11.3/worker.go:50 +0x65
```

```
s3-pod-failure-21650-etcd-0                                       1/1     Running                  0                50m     10.104.15.245   4am-node20   <none>           <none>
 s3-pod-failure-21650-etcd-1                                       1/1     Running                  0                50m     10.104.17.17    4am-node23   <none>           <none>
 s3-pod-failure-21650-etcd-2                                       1/1     Running                  0                50m     10.104.30.253   4am-node38   <none>           <none>
 s3-pod-failure-21650-milvus-datanode-8655458b49-pxnnt             1/1     Running                  2 (50m ago)      50m     10.104.24.111   4am-node29   <none>           <none>
 s3-pod-failure-21650-milvus-datanode-8655458b49-w2cld             1/1     Running                  2 (50m ago)      50m     10.104.25.181   4am-node30   <none>           <none>
 s3-pod-failure-21650-milvus-mixcoord-78c4755c9b-chspf             1/1     Running                  2 (50m ago)      50m     10.104.27.49    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-proxy-f6bdc4886-g94w5                 1/1     Running                  2 (50m ago)      50m     10.104.27.48    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-9f2g2             1/1     Running                  3 (49m ago)      50m     10.104.23.191   4am-node27   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-gcm77             1/1     Running                  2 (50m ago)      50m     10.104.27.47    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-jd559             1/1     Running                  3 (49m ago)      50m     10.104.18.184   4am-node25   <none>           <none>
 s3-pod-failure-21650-milvus-streamingnode-67b779fc-f645q          0/1     Error                    11 (3m44s ago)   50m     10.104.24.110   4am-node29   <none>           <none>
 s3-pod-failure-21650-milvus-streamingnode-67b779fc-q9sxh          1/1     Running                  11 (4m10s ago)   50m     10.104.9.55     4am-node14   <none>           <none>
 s3-pod-failure-21650-minio-0                                      1/1     Running                  8 (26m ago)      50m     10.104.15.246   4am-node20   <none>           <none>
 s3-pod-failure-21650-minio-1                                      1/1     Running                  8 (26m ago)      50m     10.104.17.15    4am-node23   <none>           <none>
 s3-pod-failure-21650-minio-2                                      1/1     Running                  8 (26m ago)      50m     10.104.30.2     4am-node38   <none>           <none>
 s3-pod-failure-21650-minio-3                                      1/1     Running                  8 (26m ago)      50m     10.104.21.45    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-0                            1/1     Running                  0                50m     10.104.15.247   4am-node20   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-1                            1/1     Running                  0                50m     10.104.17.16    4am-node23   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-2                            1/1     Running                  0                50m     10.104.21.46    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-init-nk9zh                   0/1     Completed                0                50m     10.104.13.252   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-broker-0                            1/1     Running                  0                50m     10.104.13.253   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-broker-1                            1/1     Running                  0                50m     10.104.14.123   4am-node18   <none>           <none>
 s3-pod-failure-21650-pulsarv3-proxy-0                             1/1     Running                  0                50m     10.104.13.251   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-proxy-1                             1/1     Running                  0                50m     10.104.14.122   4am-node18   <none>           <none>
 s3-pod-failure-21650-pulsarv3-pulsar-init-br57d                   0/1     Completed                0                50m     10.104.13.249   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-recovery-0                          1/1     Running                  0                50m     10.104.13.250   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-0                         1/1     Running                  0                50m     10.104.30.252   4am-node38   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-1                         1/1     Running                  0                50m     10.104.21.44    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-2                         1/1     Running                  0                50m     10.104.17.19    4am-node23   <none>           <none>
```



### Expected Behavior

_No response_

### Steps To Reproduce

```markdown

```

### Milvus Log

failed job: https://qa-jenkins.milvus.io/blue/organizations/jenkins/chaos-test-cron/detail/chaos-test-cron/21650/pipeline
log: 

[artifacts-s3-pod-failure-21650-server-logs.tar.gz](https://github.com/user-attachments/files/21012608/artifacts-s3-pod-failure-21650-server-logs.tar.gz)


cluster: 4am
ns: chaos-testing
pods

```
 s3-pod-failure-21650-etcd-0                                       1/1     Running                  0                50m     10.104.15.245   4am-node20   <none>           <none>
 s3-pod-failure-21650-etcd-1                                       1/1     Running                  0                50m     10.104.17.17    4am-node23   <none>           <none>
 s3-pod-failure-21650-etcd-2                                       1/1     Running                  0                50m     10.104.30.253   4am-node38   <none>           <none>
 s3-pod-failure-21650-milvus-datanode-8655458b49-pxnnt             1/1     Running                  2 (50m ago)      50m     10.104.24.111   4am-node29   <none>           <none>
 s3-pod-failure-21650-milvus-datanode-8655458b49-w2cld             1/1     Running                  2 (50m ago)      50m     10.104.25.181   4am-node30   <none>           <none>
 s3-pod-failure-21650-milvus-mixcoord-78c4755c9b-chspf             1/1     Running                  2 (50m ago)      50m     10.104.27.49    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-proxy-f6bdc4886-g94w5                 1/1     Running                  2 (50m ago)      50m     10.104.27.48    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-9f2g2             1/1     Running                  3 (49m ago)      50m     10.104.23.191   4am-node27   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-gcm77             1/1     Running                  2 (50m ago)      50m     10.104.27.47    4am-node31   <none>           <none>
 s3-pod-failure-21650-milvus-querynode-bb7497c76-jd559             1/1     Running                  3 (49m ago)      50m     10.104.18.184   4am-node25   <none>           <none>
 s3-pod-failure-21650-milvus-streamingnode-67b779fc-f645q          0/1     Error                    11 (3m44s ago)   50m     10.104.24.110   4am-node29   <none>           <none>
 s3-pod-failure-21650-milvus-streamingnode-67b779fc-q9sxh          1/1     Running                  11 (4m10s ago)   50m     10.104.9.55     4am-node14   <none>           <none>
 s3-pod-failure-21650-minio-0                                      1/1     Running                  8 (26m ago)      50m     10.104.15.246   4am-node20   <none>           <none>
 s3-pod-failure-21650-minio-1                                      1/1     Running                  8 (26m ago)      50m     10.104.17.15    4am-node23   <none>           <none>
 s3-pod-failure-21650-minio-2                                      1/1     Running                  8 (26m ago)      50m     10.104.30.2     4am-node38   <none>           <none>
 s3-pod-failure-21650-minio-3                                      1/1     Running                  8 (26m ago)      50m     10.104.21.45    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-0                            1/1     Running                  0                50m     10.104.15.247   4am-node20   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-1                            1/1     Running                  0                50m     10.104.17.16    4am-node23   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-2                            1/1     Running                  0                50m     10.104.21.46    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-bookie-init-nk9zh                   0/1     Completed                0                50m     10.104.13.252   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-broker-0                            1/1     Running                  0                50m     10.104.13.253   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-broker-1                            1/1     Running                  0                50m     10.104.14.123   4am-node18   <none>           <none>
 s3-pod-failure-21650-pulsarv3-proxy-0                             1/1     Running                  0                50m     10.104.13.251   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-proxy-1                             1/1     Running                  0                50m     10.104.14.122   4am-node18   <none>           <none>
 s3-pod-failure-21650-pulsarv3-pulsar-init-br57d                   0/1     Completed                0                50m     10.104.13.249   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-recovery-0                          1/1     Running                  0                50m     10.104.13.250   4am-node16   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-0                         1/1     Running                  0                50m     10.104.30.252   4am-node38   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-1                         1/1     Running                  0                50m     10.104.21.44    4am-node24   <none>           <none>
 s3-pod-failure-21650-pulsarv3-zookeeper-2                         1/1     Running                  0                50m     10.104.17.19    4am-node23   <none>           <none>
```

### Anything else?

_No response_

--- a/internal/flushcommon/syncmgr/pack_writer.go
+++ b/internal/flushcommon/syncmgr/pack_writer.go
@@ -272,7 +272,7 @@ func (bw *BulkPackWriter) writeBM25Stasts(ctx context.Context, pack *SyncPack) (
 
 	if pack.isFlush {
 		if pack.level != datapb.SegmentLevel_L0 {
-			if hasBM25Function(bw.metaCache.GetSchema(pack.tsFrom)) {
+			if hasBM25Function(bw.schema) {
 				mergedBM25Blob, err := serializer.serializeMergedBM25Stats(pack)
 				if err != nil {
 					return nil, err

--- a/internal/flushcommon/syncmgr/pack_writer_v2.go
+++ b/internal/flushcommon/syncmgr/pack_writer_v2.go
@@ -179,7 +179,8 @@ func (bw *BulkPackWriterV2) serializeBinlog(ctx context.Context, pack *SyncPack)
 	if len(pack.insertData) == 0 {
 		return nil, nil
 	}
-	arrowSchema, err := storage.ConvertToArrowSchema(bw.schema.Fields)
+
+	arrowSchema, err := storage.ConvertToArrowSchema(bw.schema.GetFields())
 	if err != nil {
 		return nil, err
 	}

--- a/internal/flushcommon/writebuffer/write_buffer_test.go
+++ b/internal/flushcommon/writebuffer/write_buffer_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/suite"
 
@@ -15,6 +16,7 @@ import (
 	"github.com/milvus-io/milvus/internal/flushcommon/metacache/pkoracle"
 	"github.com/milvus-io/milvus/internal/flushcommon/syncmgr"
 	"github.com/milvus-io/milvus/pkg/v2/common"
+	"github.com/milvus-io/milvus/pkg/v2/mq/msgstream"
 	"github.com/milvus-io/milvus/pkg/v2/proto/datapb"
 	"github.com/milvus-io/milvus/pkg/v2/util/conc"
 	"github.com/milvus-io/milvus/pkg/v2/util/paramtable"
@@ -303,4 +305,151 @@ func (s *WriteBufferSuite) TestDropPartitions() {
 
 func TestWriteBufferBase(t *testing.T) {
 	suite.Run(t, new(WriteBufferSuite))
+}
+
+// TestPrepareInsertWithMissingFields tests the missing field filling functionality
+func TestPrepareInsertWithMissingFields(t *testing.T) {
+	// Create a collection schema with multiple fields including nullable and default value fields
+	collSchema := &schemapb.CollectionSchema{
+		Name: "test_collection",
+		Fields: []*schemapb.FieldSchema{
+			{FieldID: common.RowIDField, DataType: schemapb.DataType_Int64, Name: "RowID"},
+			{FieldID: common.TimeStampField, DataType: schemapb.DataType_Int64, Name: "Timestamp"},
+			{
+				FieldID:      100,
+				Name:         "pk",
+				DataType:     schemapb.DataType_Int64,
+				IsPrimaryKey: true,
+				AutoID:       true,
+			},
+			{
+				FieldID:  101,
+				Name:     "int32_field",
+				DataType: schemapb.DataType_Int32,
+				Nullable: false,
+				DefaultValue: &schemapb.ValueField{
+					Data: &schemapb.ValueField_IntData{
+						IntData: 42,
+					},
+				},
+			},
+			{
+				FieldID:  102,
+				Name:     "nullable_float_field",
+				DataType: schemapb.DataType_Float,
+				Nullable: true,
+			},
+			{
+				FieldID:  103,
+				Name:     "default_string_field",
+				DataType: schemapb.DataType_String,
+				Nullable: false,
+				DefaultValue: &schemapb.ValueField{
+					Data: &schemapb.ValueField_StringData{
+						StringData: "default_string",
+					},
+				},
+			},
+		},
+	}
+
+	pkField := &schemapb.FieldSchema{
+		FieldID:      100,
+		Name:         "pk",
+		DataType:     schemapb.DataType_Int64,
+		IsPrimaryKey: true,
+		AutoID:       true,
+	}
+
+	t.Run("test_missing_fields_with_default_values", func(t *testing.T) {
+		// Create insert message with only some fields
+		insertMsg := &msgstream.InsertMsg{
+			BaseMsg: msgstream.BaseMsg{
+				BeginTimestamp: 1000,
+				EndTimestamp:   1002,
+			},
+			InsertRequest: &msgpb.InsertRequest{
+				SegmentID:    1,
+				PartitionID:  1,
+				CollectionID: 1,
+				FieldsData: []*schemapb.FieldData{
+					{
+						FieldId:   common.RowIDField,
+						FieldName: "RowID",
+						Type:      schemapb.DataType_Int64,
+						Field: &schemapb.FieldData_Scalars{
+							Scalars: &schemapb.ScalarField{
+								Data: &schemapb.ScalarField_LongData{
+									LongData: &schemapb.LongArray{
+										Data: []int64{1, 2, 3},
+									},
+								},
+							},
+						},
+					},
+					{
+						FieldId:   common.TimeStampField,
+						FieldName: "Timestamp",
+						Type:      schemapb.DataType_Int64,
+						Field: &schemapb.FieldData_Scalars{
+							Scalars: &schemapb.ScalarField{
+								Data: &schemapb.ScalarField_LongData{
+									LongData: &schemapb.LongArray{
+										Data: []int64{1000, 1001, 1002},
+									},
+								},
+							},
+						},
+					},
+					{
+						FieldId:   100, // Primary key field
+						FieldName: "pk",
+						Type:      schemapb.DataType_Int64,
+						Field: &schemapb.FieldData_Scalars{
+							Scalars: &schemapb.ScalarField{
+								Data: &schemapb.ScalarField_LongData{
+									LongData: &schemapb.LongArray{
+										Data: []int64{1, 2, 3},
+									},
+								},
+							},
+						},
+					},
+					{
+						FieldId:   101, // field1 with default value
+						FieldName: "int32_field",
+						Type:      schemapb.DataType_Int32,
+						Field: &schemapb.FieldData_Scalars{
+							Scalars: &schemapb.ScalarField{
+								Data: &schemapb.ScalarField_IntData{
+									IntData: &schemapb.IntArray{
+										Data: []int32{1, 2, 3},
+									},
+								},
+							},
+						},
+					},
+				},
+				// Missing nullable_float_field (nullable)
+				// Missing default_string_field (with default value)
+				NumRows:    3,
+				Version:    msgpb.InsertDataVersion_ColumnBased,
+				Timestamps: []uint64{1, 1, 1},
+			},
+		}
+
+		insertMsgs := []*msgstream.InsertMsg{insertMsg}
+		result, err := PrepareInsert(collSchema, pkField, insertMsgs)
+
+		assert.NoError(t, err)
+		assert.Len(t, result, 1)
+		assert.Equal(t, int64(3), result[0].rowNum)
+
+		// Check that all fields are present in the result
+		insertData := result[0].data[0]
+		for i := 0; i < 3; i++ {
+			assert.Nil(t, insertData.Data[102].GetRow(i))
+			assert.Equal(t, "default_string", insertData.Data[103].GetRow(i))
+		}
+	})
 }

--- a/internal/storage/serde.go
+++ b/internal/storage/serde.go
@@ -887,17 +887,24 @@ func BuildRecord(b *array.RecordBuilder, data *InsertData, fields []*schemapb.Fi
 	if data == nil {
 		return nil
 	}
+
 	for i, field := range fields {
 		fBuilder := b.Field(i)
 		typeEntry, ok := serdeMap[field.DataType]
 		if !ok {
 			panic("unknown type")
 		}
-		if data.Data[field.FieldID].RowNum() == 0 {
+		fieldData, exists := data.Data[field.FieldID]
+		if !exists {
+			return merr.WrapErrFieldNotFound(field.FieldID, fmt.Sprintf("field %s not found", field.Name))
+		}
+
+		if fieldData.RowNum() == 0 {
 			return merr.WrapErrServiceInternal(fmt.Sprintf("row num is 0 for field %s", field.Name))
 		}
-		for j := 0; j < data.Data[field.FieldID].RowNum(); j++ {
-			ok = typeEntry.serialize(fBuilder, data.Data[field.FieldID].GetRow(j))
+
+		for j := 0; j < fieldData.RowNum(); j++ {
+			ok = typeEntry.serialize(fBuilder, fieldData.GetRow(j))
 			if !ok {
 				return merr.WrapErrServiceInternal(fmt.Sprintf("serialize error on type %s", field.DataType.String()))
 			}

--- a/internal/storage/serde_events.go
+++ b/internal/storage/serde_events.go
@@ -267,7 +267,7 @@ func ValueDeserializer(r Record, v []*Value, fieldSchema []*schemapb.FieldSchema
 			dt := f.DataType
 			if r.Column(j).IsNull(i) {
 				if f.GetDefaultValue() != nil {
-					m[j] = getDefaultValue(f)
+					m[j] = GetDefaultValue(f)
 				} else {
 					m[j] = nil
 				}

--- a/internal/storage/utils.go
+++ b/internal/storage/utils.go
@@ -534,10 +534,10 @@ func ColumnBasedInsertMsgToInsertData(msg *msgstream.InsertMsg, collSchema *sche
 
 		srcField, ok := srcFields[field.GetFieldID()]
 		if !ok && field.GetFieldID() >= common.StartOfUserFieldID {
-			if !field.GetNullable() {
-				return nil, merr.WrapErrFieldNotFound(field.GetFieldID(), fmt.Sprintf("field %s not found when converting insert msg to insert data", field.GetName()))
+			err := fillMissingFields(collSchema, idata)
+			if err != nil {
+				return nil, err
 			}
-			log.Warn("insert msg missing field but nullable", zap.Int64("fieldID", field.GetFieldID()), zap.String("fieldName", field.GetName()))
 			continue
 		}
 		var fieldData FieldData
@@ -1440,7 +1440,7 @@ func IsBM25FunctionOutputField(field *schemapb.FieldSchema, collSchema *schemapb
 	return false
 }
 
-func getDefaultValue(fieldSchema *schemapb.FieldSchema) interface{} {
+func GetDefaultValue(fieldSchema *schemapb.FieldSchema) interface{} {
 	switch fieldSchema.DataType {
 	case schemapb.DataType_Bool:
 		return fieldSchema.GetDefaultValue().GetBoolData()
@@ -1462,4 +1462,46 @@ func getDefaultValue(fieldSchema *schemapb.FieldSchema) interface{} {
 		// won't happen
 		panic(fmt.Sprintf("undefined data type:%s", fieldSchema.DataType.String()))
 	}
+}
+
+// fillMissingFields fills default values or null values for missing fields in insertData
+func fillMissingFields(schema *schemapb.CollectionSchema, insertData *InsertData) error {
+	batchRows := int64(insertData.GetRowNum())
+
+	for _, field := range schema.Fields {
+		// Skip function output fields and system fields
+		if field.GetIsFunctionOutput() || field.GetFieldID() < 100 {
+			continue
+		}
+
+		_, exists := insertData.Data[field.GetFieldID()]
+
+		if !exists {
+			// Create default field data if not found
+			fieldData, err := NewFieldData(field.DataType, field, int(batchRows))
+			if err != nil {
+				return merr.WrapErrServiceInternal(fmt.Sprintf("failed to create default field data for field %s: %v", field.Name, err))
+			}
+
+			if field.GetDefaultValue() != nil { // Fill with default value
+				defaultValue := GetDefaultValue(field)
+
+				for j := 0; j < int(batchRows); j++ {
+					if err := fieldData.AppendRow(defaultValue); err != nil {
+						return merr.WrapErrServiceInternal(fmt.Sprintf("failed to append default value for field %s: %v", field.Name, err))
+					}
+				}
+			} else if field.GetNullable() { // Fill with null values
+				for j := 0; j < int(batchRows); j++ {
+					if err := fieldData.AppendRow(nil); err != nil {
+						return merr.WrapErrServiceInternal(fmt.Sprintf("failed to append null value for field %s: %v", field.Name, err))
+					}
+				}
+			} else {
+				return merr.WrapErrServiceInternal(fmt.Sprintf("field %s is not nullable and has no default value", field.Name))
+			}
+			insertData.Data[field.GetFieldID()] = fieldData
+		}
+	}
+	return nil
 }

